receiver/awscontainerinsightreceiver/internal/k8swindows/kubelet/kubelet.go

```diff
@@ -0,0 +1,206 @@
+	"fmt"
+	"os"
+	"strconv"
+
+	ci "github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/containerinsight"
+	cExtractor "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awscontainerinsightreceiver/internal/cadvisor/extractors"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awscontainerinsightreceiver/internal/k8swindows/extractors"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awscontainerinsightreceiver/internal/stores/kubeletutil"
+
+	"go.uber.org/zap"
+	stats "k8s.io/kubelet/pkg/apis/stats/v1alpha1"
+)
+
+// KubeletProvider Represents interface to kubelet.
+type KubeletProvider interface {
+	GetClient() (*kubeletutil.KubeletClient, error)
+	GetSummary() (*stats.Summary, error)
+}
+
+type kubeletProvider struct {
+	logger   *zap.Logger
+	hostIP   string
+	hostPort string
+	client   *kubeletutil.KubeletClient
+}
+
+// GetClient Returns singleton kubelet client.
+func (k *kubeletProvider) GetClient() (*kubeletutil.KubeletClient, error) {
+	if k.client != nil {
+		return k.client, nil
+	}
+	kclient, err := kubeletutil.NewKubeletClient(k.hostIP, k.hostPort, k.logger)
+	if err != nil {
+		k.logger.Error("failed to initialize kubeletProvider client, ", zap.Error(err))
+		return nil, err
+	}
+	k.client = kclient
+	return kclient, nil
+}
+
+// GetSummary Get Summary from kubelet API.
+func (k *kubeletProvider) GetSummary() (*stats.Summary, error) {
+	kclient, err := k.GetClient()
+	if err != nil {
+		k.logger.Error("failed to get kubeletProvider client, ", zap.Error(err))
+		return nil, err
+	}
+
+	summary, err := kclient.Summary(k.logger)
+	if err != nil {
+		k.logger.Error("kubeletProvider summary API failed, ", zap.Error(err))
+		return nil, err
+	}
+	return summary, nil
+}
+
+func createDefaultKubeletProvider(logger *zap.Logger) KubeletProvider {
+	return &kubeletProvider{logger: logger, hostIP: os.Getenv("HOST_IP"), hostPort: ci.KubeSecurePort}
+}
+
+// SummaryProvider represents receiver to get container metric from Kubelet.
+type SummaryProvider struct {
+	logger           *zap.Logger
+	kubeletProvider  KubeletProvider
+	hostInfo         cExtractor.CPUMemInfoProvider
+	metricExtractors []extractors.MetricExtractor
+}
+
+// Options decorates SummaryProvider struct.
+type Options func(provider *SummaryProvider)
+
+func New(logger *zap.Logger, info cExtractor.CPUMemInfoProvider, mextractor []extractors.MetricExtractor, opts ...Options) (*SummaryProvider, error) {
+	sp := &SummaryProvider{
+		logger:           logger,
+		hostInfo:         info,
+		kubeletProvider:  createDefaultKubeletProvider(logger),
+		metricExtractors: mextractor,
+	}
+
+	for _, opt := range opts {
+		opt(sp)
+	}
+	return sp, nil
+}
+
+func (k *SummaryProvider) GetMetrics() ([]*cExtractor.CAdvisorMetric, error) {
+	var metrics []*cExtractor.CAdvisorMetric
+
+	summary, err := k.kubeletProvider.GetSummary()
+	if err != nil {
+		k.logger.Error("kubeletProvider summary API failed, ", zap.Error(err))
```

Comment:
GetSummary also has the exact same error.  is there a better way to do this?

This pattern is in a few spots.  It doesn't seem useful unless each level is more specific about what failed or what it means

Reply:
Addressed, improved logging and naming for consistency